### PR TITLE
feat: add listener metrics

### PR DESCRIPTION
**Pull request by acud**
This change is [Reviewable](https://reviewable.io/reviews/ethersphere/bee/1786)

### CODE REVIEW

**Review by ralph-pichler — APPROVED** [review]

**Review by anatollupacescu — APPROVED** [review]

**Review by aloknerurkar — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1786)* status: 0 of 3 files reviewed, 1 unresolved discussion (waiting on @acud)

___
*[pkg/postage/listener/listener.go, line 230 at r1](https://reviewable.io/reviews/ethersphere/bee/1786#-Ma7kz9f4DO8-F2VUss4:-Ma7kz9f4DO8-F2VUss5:b-o3pl0o) ([raw file](https://github.com/ethersphere/bee/blob/e2d1fd33e33ff909d7cafa7595994c480d632428/pkg/postage/listener/listener.go#L230)):*
> ```Go
> 
> 			from = to + 1
> 			totalTimeMetric(l.metrics.PageProcessDuration, start)
> ```

Would it be better to have a separate PagesProcessed counter? I see BackendCalls happening twice per page. So we should be able to deduce from that. But as we have aggregated duration it would be better to understand how many of these happened to figure out the avg time.

**Comment by aloknerurkar** [comment]
___
*[pkg/postage/listener/listener.go, line 221 at r1](https://reviewable.io/reviews/ethersphere/bee/1786#-Ma7mE_SCjOCVwJqo-Xm:-Ma7kO1mDsf_tcZeaUnA:b7qihj9) ([raw file](https://github.com/ethersphere/bee/blob/e2d1fd33e33ff909d7cafa7595994c480d632428/pkg/postage/listener/listener.go#L221)):*
> ```Go
> 					return err
> 				}
> 				totalTimeMetric(l.metrics.EventProcessDuration, startEv)
> ```

In processEvent, we increment the EventsProcessed counter regardless of the error. We will have a bunch of EventsProcessed which wont have added the ProcessDuration in this case.

**Comment by aloknerurkar** [comment]
___
*[pkg/postage/listener/listener.go, line 224 at r1](https://reviewable.io/reviews/ethersphere/bee/1786#-Ma7n7E_4EYdLgdFjy_y:-Ma7n7E_4EYdLgdFjy_z:bc1je1a) ([raw file](https://github.com/ethersphere/bee/blob/e2d1fd33e33ff909d7cafa7595994c480d632428/pkg/postage/listener/listener.go#L224)):*
> ```Go
> 			}
> 
> 			err = updater.UpdateBlockNumber(to)
> ```

This can be removed no?

**Review by esadakar — APPROVED** [review]

**Review by acud — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1786)* status: 0 of 3 files reviewed, 3 unresolved discussions (waiting on @acud and @aloknerurkar)

___
*[pkg/postage/listener/listener.go, line 221 at r1](https://reviewable.io/reviews/ethersphere/bee/1786#-Ma7mE_SCjOCVwJqo-Xm:-Ma7ojf38TlkyUBF1fye:b-6r07yz) ([raw file](https://github.com/ethersphere/bee/blob/e2d1fd33e33ff909d7cafa7595994c480d632428/pkg/postage/listener/listener.go#L221)):*
<details><summary><i>Previously, aloknerurkar (Alok Nerurkar) wrote…</i></summary><blockquote>

In processEvent, we increment the EventsProcessed counter regardless of the error. We will have a bunch of EventsProcessed which wont have added the ProcessDuration in this case. 
</blockquote></details>

not sure if this is so relevant. right now, from what I gather, the listener will actually stop listening in the case that an error has occured. so there would not be any subsequent executions if there was an error
___
*[pkg/postage/listener/listener.go, line 224 at r1](https://reviewable.io/reviews/ethersphere/bee/1786#-Ma7n7E_4EYdLgdFjy_y:-Ma7oDwe1js6SRLroc_L:b-jybzvn) ([raw file](https://github.com/ethersphere/bee/blob/e2d1fd33e33ff909d7cafa7595994c480d632428/pkg/postage/listener/listener.go#L224)):*
<details><summary><i>Previously, aloknerurkar (Alok Nerurkar) wrote…</i></summary><blockquote>

This can be removed no?
</blockquote></details>

no, since there might be no events for the requested interval
___
*[pkg/postage/listener/listener.go, line 230 at r1](https://reviewable.io/reviews/ethersphere/bee/1786#-Ma7kz9f4DO8-F2VUss4:-Ma7pjysEk9nMUqty-Do:b-896fix) ([raw file](https://github.com/ethersphere/bee/blob/e2d1fd33e33ff909d7cafa7595994c480d632428/pkg/postage/listener/listener.go#L230)):*
<details><summary><i>Previously, aloknerurkar (Alok Nerurkar) wrote…</i></summary><blockquote>

Would it be better to have a separate PagesProcessed counter? I see BackendCalls happening twice per page. So we should be able to deduce from that. But as we have aggregated duration it would be better to understand how many of these happened to figure out the avg time.
</blockquote></details>

Done.

**Review by mrekucci — APPROVED** [review]
:lgtm:

> Reviewed 1 of 3 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1786#-:-Ma7ttCE2zLdrtHzC8Cd:bnfp4nl)* status: all files reviewed, 3 unresolved discussions (waiting on @aloknerurkar)

___
*[pkg/postage/listener/metrics.go, line 16 at r2](https://reviewable.io/reviews/ethersphere/bee/1786#-Ma7sTdf-AKbupwPZkNO:-Ma7sTdf-AKbupwPZkNP:b6qnm4b) ([raw file](https://github.com/ethersphere/bee/blob/04fdeb7a5b9355bc2fd5717c662d128010f21ddf/pkg/postage/listener/metrics.go#L16)):*
> ```Go
> 	// individual event counters
> 	CreatedCounter prometheus.Counter
> 	TopupCounter   prometheus.Counter
> ```

Since the event is called `"BatchTopUp"` I'd rename the `TopUpCounter` into`TopupCounter` and so all the related fields as `batchTopupTopic` and `TopupAmount` in the `batchTopUpEvent` type.

**Review by janos — APPROVED** [review]

**Review by aloknerurkar — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1786)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @acud)

File: pkg/postage/listener/metrics.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+import (
+	m "github.com/ethersphere/bee/pkg/metrics"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type metrics struct {
+	// aggregate events handled
+	EventsProcessed prometheus.Counter
+	EventErrors     prometheus.Counter
+	PagesProcessed  prometheus.Counter
```
Comment (janos) [inline pkg/postage/listener/metrics.go:12]: PagesProcessed is not constructed in newMetrics function bellow.

**Review by janos — CHANGES_REQUESTED** [review]
Just assign a missing metric in newMetrics.

**Review by mrekucci — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/ethersphere/bee/1786)* status: all files reviewed, 1 unresolved discussion (waiting on @acud)

___
*[pkg/postage/listener/metrics.go, line 107 at r2](https://reviewable.io/reviews/ethersphere/bee/1786#-Ma7wzE29vEbWhAElfw3:-Ma7wzE29vEbWhAElfw4:budjo08) ([raw file](https://github.com/ethersphere/bee/blob/04fdeb7a5b9355bc2fd5717c662d128010f21ddf/pkg/postage/listener/metrics.go#L107)):*
> ```Go
> }
> 
> func (s *listener) Metrics() []prometheus.Collector {
> ```

Just a niche, the `s` should be `l`.

**Review by janos — APPROVED** [review]

**Pull request merged**